src/mapping/query/function/mod.rs

```diff
@@ -0,0 +1,226 @@
+#![macro_use]
+
+mod not;
+pub(in crate::mapping) use not::NotFn;
+
+use super::Function;
+use crate::Event;
+use crate::{event::Value, mapping::Result};
+use std::collections::HashMap;
+use std::convert::TryFrom;
+use std::str::FromStr;
+
+// If this macro triggers, it means the logic to detect invalid types did not
+// function as expected. This is a bug in the implementation.
+macro_rules! unexpected_type {
+    ($value:expr) => {
+        unreachable!("unexpected value type: '{}'", $value.kind());
+    };
+}
+
+macro_rules! build_signatures {
+    ($($name:ident => $func:ident),* $(,)?) => {
+        $(mod $name;)*
+
+        $(pub(in crate::mapping) use self::$name::$func;)*
+
+        #[derive(Debug, Copy, Clone, Eq, PartialEq)]
+        #[allow(clippy::enum_variant_names)]
+        pub(in crate::mapping) enum FunctionSignature {
+            $($func,)*
+        }
+
+        impl FromStr for FunctionSignature {
+            type Err = String;
+
+            fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
+                let func = match s {
+                    $(stringify!($name) => Self::$func,)*
+                    _ => return Err(format!("unknown function '{}'", s)),
+                };
+
+                Ok(func)
+            }
+        }
+
+        impl FunctionSignature {
+            pub fn as_str(&self) -> &str {
+                match self {
+                    $(Self::$func => stringify!($name),)*
+                }
+            }
+
+            pub fn parameters(&self) -> &[Parameter] {
+                match self {
+                    $(Self::$func => $func::parameters(),)*
+                }
+            }
+
+            pub fn into_boxed_function(self, arguments: ArgumentList) -> Result<Box<dyn Function>> {
+                match self {
+                    $(Self::$func => $func::try_from(arguments)
+                        .map(|func| Box::new(func) as Box<dyn Function>),)*
+                }
+            }
+        }
+    };
+}
+
+// List of built-in functions.
+build_signatures! {
+    to_string => ToStringFn,
+    to_int => ToIntegerFn,
+    to_float => ToFloatFn,
+    to_bool => ToBooleanFn,
+    to_timestamp => ToTimestampFn,
+    parse_timestamp => ParseTimestampFn,
+    strip_whitespace => StripWhitespaceFn,
+    upcase => UpcaseFn,
+    downcase => DowncaseFn,
+    uuid_v4 => UuidV4Fn,
+    md5 => Md5Fn,
+    sha1 => Sha1Fn,
+    now => NowFn,
+    truncate => TruncateFn,
+    parse_json => ParseJsonFn,
+}
+
+/// Commonly used types when building new functions.
+mod prelude {
```

Comment:
Minute nitpick - I think this would be better at the top of the file.

Reply:
done